test/cpp/ext/otel/otel_plugin_test.cc

```diff
@@ -1353,6 +1411,436 @@ TEST_F(OpenTelemetryPluginNPCMetricsTest,
+                            opentelemetry::sdk::metrics::LastValuePointData>(
+                            data[kMetricName][i - 1].point_data)
+                            .value_)),
+                    GaugeResultLaterThan(
+                        opentelemetry::nostd::get<
+                            opentelemetry::sdk::metrics::LastValuePointData>(
+                            data[kMetricName][i - 1].point_data)
+                            .sample_ts_)));
+  }
+  // Verify labels.
+  EXPECT_THAT(data,
+              ::testing::UnorderedElementsAre(::testing::Pair(
+                  kMetricName, ::testing::Each(AttributesEq(
+                                   kLabelKeys, kLabelValues, kOptionalLabelKeys,
+                                   kOptionalLabelValues)))));
+}
+
+using OpenTelemetryPluginCallbackMetricsTest = OpenTelemetryPluginEnd2EndTest;
+
+// The report duration is longer than the collect duration, so we expect to
+// collect duplicated (cached) values.
+TEST_F(OpenTelemetryPluginCallbackMetricsTest,
+       ReportDurationLongerThanCollectDuration) {
+  constexpr absl::string_view kInt64CallbackGaugeMetric =
+      "int64_callback_gauge";
+  constexpr absl::string_view kDoubleCallbackGaugeMetric =
+      "double_callback_gauge";
+  constexpr std::array<absl::string_view, 2> kLabelKeys = {"label_key_1",
+                                                           "label_key_2"};
+  constexpr std::array<absl::string_view, 2> kOptionalLabelKeys = {
+      "optional_label_key_1", "optional_label_key_2"};
+  constexpr std::array<absl::string_view, 2> kLabelValues = {"label_value_1",
+                                                             "label_value_2"};
+  constexpr std::array<absl::string_view, 2> kOptionalLabelValues = {
+      "optional_label_value_1", "optional_label_value_2"};
+  auto integer_gauge_handle =
+      grpc_core::GlobalInstrumentsRegistry::RegisterCallbackInt64Gauge(
+          kInt64CallbackGaugeMetric, "An int64 callback gauge.", "unit",
+          kLabelKeys, kOptionalLabelKeys,
+          /*enable_by_default=*/true);
+  auto double_gauge_handle =
+      grpc_core::GlobalInstrumentsRegistry::RegisterCallbackDoubleGauge(
+          kDoubleCallbackGaugeMetric, "A double callback gauge.", "unit",
+          kLabelKeys, kOptionalLabelKeys,
+          /*enable_by_default=*/true);
+  Init(std::move(Options()
+                     .set_metric_names({kInt64CallbackGaugeMetric,
+                                        kDoubleCallbackGaugeMetric})
+                     .add_optional_label(kOptionalLabelKeys[0])
+                     .add_optional_label(kOptionalLabelKeys[1])));
+  auto stats_plugins =
+      grpc_core::GlobalStatsPluginRegistry::GetStatsPluginsForChannel(
+          grpc_core::StatsPlugin::ChannelScope("dns:///localhost:8080", ""));
+  int64_t int_value = 0;
+  double double_value = 0.5;
+  auto registered_metric_callback = stats_plugins.RegisterCallback(
```

Comment:
Please test with multiple callbacks for the same metric, each reporting different label values.

Reply:
Done.